### PR TITLE
Group refactor

### PR DESCRIPTION
**Pull request by gilleain**
Refactor of the group (partition refinement) machinery. Does:

- Move the logic to get connectivity from a molecule behind a `Refinable` interface
- This allows removal of the different EquitableRefiner classes and abstract base class
- This in turn simplifies the DiscreteRefiner hierarchy
- Finally, the main functional point is to fix the equitable refinement for multigraphs (like molecules)
- An `Invariant` interface provides two implementations, one for simple and one for multigraphs

### CODE REVIEW

File: tool/group/src/test/java/org/openscience/cdk/group/EquitablePartitionRefinerTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package org.openscience.cdk.group;
```
Comment (egonw) [inline tool/group/src/test/java/org/openscience/cdk/group/EquitablePartitionRefinerTest.java]: Please add the missing header.

File: tool/group/src/test/java/org/openscience/cdk/group/BondRefinableTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        String acpString = "C0C1C2C3 0:1(1),0:3(1),1:2(1),2:3(1)";
+        BondRefinable bondRefinable = refinable(acpString);
+        assertEquals(1, bondRefinable.getConnectivity(0, 1));
+        assertEquals(1, bondRefinable.getConnectivity(0, 2));
+        assertEquals(1, bondRefinable.getConnectivity(1, 3));
+        assertEquals(1, bondRefinable.getConnectivity(2, 3));
+    }
+    
+    // TODO - neighbours in block test
+//    @Test
```
Comment (egonw) [inline tool/group/src/test/java/org/openscience/cdk/group/BondRefinableTest.java]: Maybe use @Ignore("TODO - neighbours in block test") ?

File: tool/group/src/test/java/org/openscience/cdk/group/BondRefinableTest.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+import org.openscience.cdk.interfaces.IAtomContainer;
+import org.openscience.cdk.interfaces.IChemObjectBuilder;
+import org.openscience.cdk.silent.SilentChemObjectBuilder;
+
+/**
+ * @author maclean
+ * @cdk.module group
```
Comment (egonw) [inline tool/group/src/test/java/org/openscience/cdk/group/BondRefinableTest.java]: Double check with @johnmay, but I don't think we're using these anymore... (no action needed, but removing them is an option...)

File: tool/group/src/main/java/org/openscience/cdk/group/Invariant.java [hunk 4]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (C) 2012  Gilleain Torrance <gilleain.torrance@gmail.com>
+/* Copyright (C) 2017  Gilleain Torrance <gilleain.torrance@gmail.com>
```
Comment (egonw) [inline tool/group/src/main/java/org/openscience/cdk/group/Invariant.java]: Or 2012-2017 or 2012,2017 ...

File: tool/group/src/main/java/org/openscience/cdk/group/BondRefinable.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
+package org.openscience.cdk.group;
```
Comment (egonw) [inline tool/group/src/main/java/org/openscience/cdk/group/BondRefinable.java]: Please add missing header.

**Comment by johnmay** [comment]
Yep will look into this at the weekend. Reminds of the n things I need to do also, I did external (in beam) work on canonical labelling myself. It would be good to integrate this all under one umbrella at some point.

- saucy-like partition sub-linear partition refinement, with this you can do Titin (~350k atoms) in ~100 ms (see slide 17 https://www.slideshare.net/NextMoveSoftware/cinf-1-generating-canonical-identifiers-for-glycoproteins-and-other-chemically-modified-biopolymers).
- McKay's Lemma 2.25 is really useful, asked him for the proof that I still need to understand
- stereochemistry refinement/leaf comparison handling
- randomized permutation group (since I wanted canon no need to be complete)

File: tool/group/src/test/java/org/openscience/cdk/group/BondRefinableTest.java (hunk 3 above)
Comment (gilleain) [inline tool/group/src/test/java/org/openscience/cdk/group/BondRefinableTest.java]: Had forgotten about this - happy to remove them all, but haven't yet. Other changes made.

**Comment by johnmay** [comment]
Would it be possible to make the public classes package-private and encapsulate as much as possible behind a stable public API? Code looks fine but I don't like the API breakages, we can minimise this in future if the classes are package-private.

**Comment by gilleain** [comment]
Ok, given it a go. Now split into public interfaces, and a factory for creating refiners:

`AtomContainerDiscretePartitionRefiner refiner = PartitionRefinement.forAtoms().create();`

which is not exactly wonderful, but I couldn't think of a better mechanism.

**Comment by johnmay** [comment]
All good thanks for patches - please try to keep public method signatures intact in future but otherwise okay. 

I'll remove the print to standard out (AtomContainerPrinter) from the Azulene test.

**Pull request merged**